test/functional/homepage_controller_test.rb

```diff
@@ -104,13 +104,84 @@ class HomepageControllerTest < ActionController::TestCase
       assert_redirected_to show_popular_links_path
     end
 
-    should "render edit template on errors" do
+    should "render edit template on validation error" do
+      links_with_missing_items = { "1" => { "title" => "title has changed", "url" => "https://www.url1.com" } }
+
       patch :update, params: { id: @popular_links.id,
-                               "popular_links" =>
-                                 { "1" => { "title" => "title has changed", "url" => "https://www.url1.com" } } }
+                               "popular_links" => links_with_missing_items }
 
       assert_template "homepage/popular_links/edit"
     end
+
+    context "database errors" do
+      setup do
+        PopularLinksEdition.any_instance.stubs(:save).raises(Mongoid::Errors::MongoidError.new)
+      end
+
+      should "alert 'application error'" do
+        new_title = "title has changed"
+
+        patch :update, params: { id: @popular_links.id,
+                                 "popular_links" =>
+                                   { "1" => { "title" => new_title, "url" => "https://www.url1.com" },
+                                     "2" => { "title" => "title2", "url" => "https://www.url2.com" },
+                                     "3" => { "title" => "title3", "url" => "https://www.url3.com" },
+                                     "4" => { "title" => "title4", "url" => "https://www.url4.com" },
+                                     "5" => { "title" => "title5", "url" => "https://www.url5.com" },
+                                     "6" => { "title" => "title6", "url" => "https://www.url6.com" } } }
+
+        assert_equal "Due to an application error, the edition couldn't be saved.", flash[:danger]
+      end
+
+      should "render edit template" do
+        new_title = "title has changed"
+
+        patch :update, params: { id: @popular_links.id,
+                                 "popular_links" =>
+                                   { "1" => { "title" => new_title, "url" => "https://www.url1.com" },
+                                     "2" => { "title" => "title2", "url" => "https://www.url2.com" },
+                                     "3" => { "title" => "title3", "url" => "https://www.url3.com" },
+                                     "4" => { "title" => "title4", "url" => "https://www.url4.com" },
+                                     "5" => { "title" => "title5", "url" => "https://www.url5.com" },
+                                     "6" => { "title" => "title6", "url" => "https://www.url6.com" } } }
+
+        assert_template "homepage/popular_links/edit"
+      end
+    end
+
+    context "publishing api errors" do
+      should "render edit template on publishing api error" do
+        new_title = "title has changed"
+        stub_publishing_api_put_content_error
```

Comment:
Since we've got a setup method in the previous context block (for database errors), we should probably be consistent and do the same here (for stubbing the publishing api).

Reply:
I did think about it but there are different stubbing for publishing api inside this context block and hence chose not to use setup block